### ISSUE TITLE
Bug: set_price() doesn't work when using add_to_cart in 3.5.0

### ISSUE THREAD

**Issue by indextwo** [woocommerce/woocommerce#21649]
**Describe the bug**

We have a very unique setup with a lot of conditional requirements, but: when using `add_to_cart()` dynamically to add a product to the cart, then using the `woocommerce_before_calculate_totals`  action to change the price, `set_price()` no longer seems to work as of 3.5.0. Downgrading to 3.4.7, functionality works as expected.

**To Reproduce**

### 1. Create an ajax function to add an item to the cart:

	function my_productAddToBasket() {
		
		//	Add our 'main' product
		
		WC()->cart->add_to_cart(123, 1, 0, array(), array('main-item' => true));
		
		//	Add our free product
		
		WC()->cart->add_to_cart(321, 1, 0, array(), array('free-gift' => true));
		
		$jsonArray['response'] = 'success';
		$jsonArray['cart-items'] = WC()->cart->cart_contents_count;
		wp_send_json($jsonArray);
	}

	add_action('wp_ajax_productAddToBasket', 'my_productAddToBasket');

### 2. Add an action to change the price of the `free-item`

	function vnmWooExtend_freeGiftItem_zeroPrice($cart_object) {
		$customPrice = 0;
		
		foreach ($cart_object->get_cart() as $value) {
			
			//	If it's a free gift (which is determined when being added), then make it free!
			
			if ($value['free-gift']) {
				$value['data']->set_price($customPrice);
			}
		}
	}

	add_action('woocommerce_before_calculate_totals', 'vnmWooExtend_freeGiftItem_zeroPrice');

**Expected behavior**
In all prior versions of WC up to `3.4.7`, this worked as expected - the free gift was automatically added, and its price was set to 0. Since the update to `3.5.0`, the `free-gift` item's price is no longer changed. All the conditions are met (as confirmed by adding `error_log()` lines at each step, and it knows that it _should_ change the price to 0... but it never works. 

Downgrading WC installation to `3.4.7` allows this to work as expected.

**Comment by kloon** [comment]
Can you confirm if you get the same results when using set_regular_price? Price is just a display field, to change the actual price value you need to change either the regular_price or sale_price values.

**Comment by indextwo** [comment]
> Can you confirm if you get the same results when using set_regular_price?

Confirmed. I updated my filter to:
	function vnmWooExtend_freeGiftItem_zeroPrice($cart_object) {
		$customPrice = 0;
		
		foreach ($cart_object->get_cart() as $value) {
			if ($value['free-gift']) {
				$value['data']->set_regular_price($customPrice);
			}
		}
	}

	add_action('woocommerce_before_calculate_totals', 'vnmWooExtend_freeGiftItem_zeroPrice');

...but the same result. The item that's supposed to be free is still full price, and the cart total reflects that too.

**Comment by indextwo** [comment]
^ In fact, just to confirm: `set_regular_price()` against a `['data']` item in the cart doesn't work even in WC 3.4.7 - only `set_price()` works.

**Comment by kloon** [comment]
Try using the `woocommerce_add_cart_item` and `woocommerce_get_cart_item_from_session` hooks instead. I tested these hooks on 3.5 and they work for me.

Did you confirm that the `woocommerce_before_calculate_totals` hook is actually firing for you?

**Comment by kloon** [comment]
@indextwo OK, if it is firing it should work then. My only other suggestion would be to alter the priority of the hook, try a higher priority like 90 to ensure it runs after any other functions hooked onto that. I presume you have also tested this with all plugins except WooCommerce disabled and with a default theme like Storefront or Twentyseventeen?

**Comment by indextwo** [comment]
I did try it with all plugins off except the one that had my ajax hooks and WC, but I hadn't tried switching themes (as it had all my fancy JS for doing the ajax-adds). I switched to TwentySeventeen and navigated directly to `/wp-admin/admin-ajax.php?action=productAddToBasket` and... totally worked. This was an error on my part, as I didn't look closely enough: the _item_ total was wrong (showed £21.98 when it should have been £0), but the _cart_ total was correct (75.00 for both items rather than 96.98). If I jump straight to the checkout, the item total appears as £0 as expected.

It looks like there's something in the returned fragments that's not updating the visual item totals correctly or is somehow cached (or triggered before `set_price()` is called now). Going back to my original build (with everything enabled again), when I add the free item to the cart, it _does_ add it as free, but with WC 3.5.0, when `WC_AJAX::get_refreshed_fragments()` is triggered by a custom ajax call, the fragment it returns has the original item price, even though it has clearly been added as free. Again, with WC 3.4.7 this doesn't happen. 

Has the priority with fragments changed, or is there some additional caching going on there?

**Comment by kloon** [comment]
Have you tried using a later priority as I suggested? All I can think of is that your code if firing before WooCommerce core code and this is resulting in the price being overwritten when calculating the cart object.

I did tests and I am unable to replicate, I did found that the mini cart was delayed when I did not use a later priority, bumping it to 20 made it always display correctly.

Code I used to test this:
```
function test_woocommerce_before_calculate_totals( $cart ) {
	foreach ( $cart->get_cart() as $cart_item ) {
		$cart_item['data']->set_price( 0 );
	}
}
add_action( 'woocommerce_before_calculate_totals', 'test_woocommerce_before_calculate_totals', 20 );
```

**Comment by kloon** [comment]
Closing this as I am unable to reproduce and believe the issue is with your custom code. Will gladly reopen if you can replicate this constantly with just WooCommerce, a default theme and the code example I provided.

**Comment by indextwo** [comment]
I think you missed all of my last message: 

##TL;DR:

I was wrong: `woocommerce_before_calculate_totals'` is _not_ the issue. This does set the price to 0 as expected. The problem is with this line in the mini-cart:

    $product_price = apply_filters('woocommerce_cart_item_price', WC()->cart->get_product_price($_product), $cart_item, $cart_item_key);

As illustrated in my screenshots above and confirmed on a clean install (just WooCommerce, StoreFront, an a 19-line plugin to change the price via the aforementioned hook): the price has been changed, and is confirmed as being changed if you view the actual basket, but the **mini-cart** still displays the individual item price as full; yet it still displays the cart total correctly.

As I said, if this should be opened as a new ticket, let me know. But this is definitely a bug as far as I can determine.

**Comment by kloon** [comment]
@indextwo refresh the page a couple of times and you will see the mini-cart update to the correct price. As per previous comments, change the priority in your custom code, it fixes the issue. I gave you an example.

Your code is firing too late causing the mini-cart to show the old price, on a refresh, it catches up and shows the right price.

**Comment by indextwo** [comment]
Refreshing the page doesn't make a difference. Nor does changing the priority, either to a higher or lower value (anywhere from 1 to 99999). 

The fact that the price is changed isn't the issue (as I mistakenly thought): it's that the price call in the mini-cart is wrong. Even if it's force-refreshed with another regular product added to the cart, it doesn't update.

**Comment by therealgilles** [comment]
I think I am having this same issue.

**Comment by ostinlviv** [comment]
I have same issue

**Comment by Collie-IT** [comment]
Same Issue here.

**Comment by ultrawebmarketing** [comment]
I think the issue is with what is the correct way to get the price for an item in the cart?

The official documentation states to use get_price():

https://woocommerce.wp-a2z.org/oik_api/wc_productget_price/

However the mini cart is using get_product_price():

https://woocommerce.wp-a2z.org/oik_api/wc_cartget_product_price/

So whatever method is the correct method to get the cart item's price is the one that should be used and the other method should deprecated to prevent this confusion in the future.

**Comment by Bradzer** [comment]
This issue does not seem to be fixed. Is there any workaround to solve it ?

**Comment by anothercreative** [comment]
**+1**

We are experiencing the same issue.

**Comment by katsar0v** [comment]
`set_price()` still does not work for me inside `woocommerce_before_calculate_totals` ... please re-open the issue

**Comment by CrazySquareRoot** [comment]
Well, I'm not surprised such bugs exists if so called cart items hold objects of product-definition classes. It's insane. These should be completely separate entities, otherwise any customization in cart item done through product object may well result in changed product definitions in case of accidental save of this object (and vice versa - in case of accidental read). This in and out of itself should be considered a very dangerous bug on an architectural level.

**Comment by s1awek** [comment]
For me, the problem was a mini cart - it didn't refresh after `set_price()`. I think it my by happening because of combination AJAX and caching. For example removing even one item from mini cart refreshed it and prices magically jumped in as they should.
I found this bit of code, and it's working for me:

```
// Mini cart: Display custom price 
add_filter('woocommerce_cart_item_price', 'filter_cart_item_price', 199999, 3);
function filter_cart_item_price($price_html, $cart_item, $cart_item_key)
{

	if (isset($cart_item['discounted_price'])) {
		$args = array('price' => 0);

		if (WC()->cart->display_prices_including_tax()) {
			$product_price = wc_get_price_including_tax($cart_item['data'], $args);
		} else {
			$product_price = wc_get_price_excluding_tax($cart_item['data'], $args);
		}
		return wc_price($product_price);
	}
	return $price_html;
}
```

**Comment by adamkentdev** [comment]
This approach is not possible with simple products, but is with variations. I believe this is because of the way variations are instantiated underneath the hood. Please note that $cart_item['data'] is = to the $product object.

```
        $variationProduct = wc_get_product( 5040 );
        $simpleProduct = wc_get_product( 4626 );

        echo ('type ' . $simpleProduct->get_type()  .' old price ' . $simpleProduct->get_price());
        echo ('type ' . $variationProduct->get_type()  .' old price ' . $variationProduct->get_price());

        $variationProduct->set_price( 5.00 );
        $simpleProduct->set_price( 5.00 );

        echo('type ' . $simpleProduct->get_type()  .' new price ' . $simpleProduct->get_price());
        echo('type ' . $variationProduct->get_type()  .' new price ' . $variationProduct->get_price());
```

outputs as:

type simple old price 43.25type variation old price 13.38
type simple new price 43.25 type variation new price 5


The real problem is that it doesn't seem possible to override cart line item prices in any other way prior to total calculations:

WC_Cart_Totals->calculate_item_subtotals() sets $item->subtotal = $item->price, which is the things that is effectively read only.

It's possible to override the lineitem sub totals after total calculation, but unfortunately calling WC()->cart->calculate_totals(); then recalcs the line item subtotals from the database prices again.

@mikejolley can I add a filter to  WC_Cart_Totals->calculate_item_subtotals()

```
$item->subtotal = apply_filters( 'woocommerce_adjust_cart_line_item_subtotal', $item->price, $item);
```

This will significantly improve the workflow.

Version 10.4.2
